sound/soc/sof/virtio-fe.c

```diff
@@ -19,112 +20,428 @@
+
+	virtqueue_kick(vfe->ipc_cmd_tx_vq);
+
+	return ret;
 }
 
-/*
- * IPC Mailbox IO
- */
+/* get IPC reply from BE */
+static int vfe_get_reply(struct snd_sof_dev *sdev,
+			 struct snd_sof_ipc_msg *msg)
+{
+	struct sof_vfe *vfe = sdev->vfe;
+
+	/* is reply ready ? */
 
-static void virtio_fe_mailbox_write(struct snd_sof_dev *sdev, u32 offset,
-				    void *message, size_t bytes)
+	vfe->msg = NULL;
+	return 0;
+}
+
+/* get stream message from virtio */
+static int vfe_get_stream_message(struct snd_sof_dev *sdev)
 {
-	/* write data to message Q buffer before sending message */
+	struct sof_vfe *vfe = sdev->vfe;
+	void *buf = NULL;
+	unsigned int buflen = 0;
+
+	buf = virtqueue_get_buf(vfe->ipc_not_rx_vq, &buflen);
+	if (unlikely(!buf)) {
+		dev_err(sdev->dev, "error rx not from virtio:%d!\n", buflen);
+		return -ENOMEM; /* TODO do we need to free buf ??? */
+	}
+
+	/* copy notification to IPC buffer */
+	/* memmove(msg->reply_data, buf, msg->reply_size); */
+
+	return 0;
 }
 
-static void virtio_fe_mailbox_read(struct snd_sof_dev *sdev, u32 offset,
-				   void *message, size_t bytes)
+/* tell DSP we have processed notification */
+static int vfe_cmd_done(struct snd_sof_dev *sdev, int dir)
 {
-	/* read data from message Q buffer after receiving message */
+	return 0;
 }
 
-static int virtio_fe_tx_busy(struct snd_sof_dev *sdev)
+/* send the IPC message completed, this means the BE has received the cmd */
+static void vfe_cmd_tx_done(struct virtqueue *vq)
 {
-	/* return 1 if tx message Q is busy */
+	struct snd_sof_ipc_msg *msg;
+	struct sof_vfe *vfe;
+
+	vfe = vq->vdev->priv;
+	msg = vfe->msg;
+	msg->complete = true;
+	wake_up(&msg->waitq);
 }
 
-static int virtio_fe_tx_msg(struct snd_sof_dev *sdev,
-			    struct snd_sof_ipc_msg *msg)
+static void vfe_cmd_handle_rx(struct virtqueue *vq)
 {
-	/* write msg to the virtio queue message for BE */
+}
 
-	return 0;
+static void vfe_not_tx_done(struct virtqueue *vq)
+{
 }
 
-static int virtio_fe_rx_msg(struct snd_sof_dev *sdev,
-			    struct snd_sof_ipc_msg *msg)
+static void vfe_posn_update(struct work_struct *work)
 {
-	/* read the virtio queue message from BE and copy to msg */
-	return 0;
+	struct sof_ipc_stream_posn *posn = NULL;
+	struct snd_sof_pcm *spcm;
+	struct scatterlist sg;
+	struct snd_sof_dev *sdev;
+	struct virtqueue *vq;
+	unsigned int buflen = 0;
+	int direction;
+	struct sof_vfe *vfe =
+		container_of(work, struct sof_vfe, posn_update_work);
+
+	vq = vfe->ipc_not_rx_vq;
+	sdev = vfe->sdev;
+
+	/* no lock is necessary here */
+	while ((posn = virtqueue_get_buf(vq, &buflen)) != NULL) {
+		spcm = snd_sof_find_spcm_comp(sdev, posn->comp_id, &direction);
+		if (!spcm) {
+			dev_err(sdev->dev,
+				"err: period elapsed for unused component %d\n",
+					posn->comp_id);
+			/* kick back the empty posn buffer immediately */
+			sg_init_one(&sg, posn, sizeof(*posn));
+			virtqueue_add_inbuf(vq, &sg, 1, posn, GFP_KERNEL);
+			virtqueue_kick(vq);
+			return;
```

Comment:
should we really return here ? what happens if we have > 1 stream and it has position messages in the queue ?